gpytorch/models/exact_prediction_strategies.py

```diff
@@ -196,27 +199,29 @@ def get_fantasy_strategy(self, inputs, targets, full_inputs, full_targets, full_
 
         lower_left = fant_train_covar.matmul(L_inverse)
         schur_root = torch.cholesky(fant_fant_covar - lower_left.matmul(lower_left.transpose(-2, -1)))
-        upper_right = torch.zeros(m, schur_root.size(-1), device=L.device, dtype=L.dtype)
 
         # Form new root Z = [L 0; lower_left schur_root]
         num_fant = schur_root.size(-2)
         m, n = L.shape[-2:]
         new_root = torch.zeros(*batch_shape, m + num_fant, n + num_fant, device=L.device, dtype=L.dtype)
         new_root[..., :m, :n] = L
-        new_root[..., :m, n:] = upper_right
         new_root[..., m:, :n] = lower_left
         new_root[..., m:, n:] = schur_root
 
         # Use pseudo-inverse of Z as new inv root
-        # TODO: Replace pseudo-inverse calculation with something more stable than normal equations once
-        # one of torch.svd, torch.qr, or torch.pinverse works in batch mode.
-        cap_mat = new_root.transpose(-2, -1).matmul(new_root)
-        if cap_mat.requires_grad or new_root.requires_grad:
+        # TODO: Consider replacing pseudo-inverse calculation with torch.svd / torch.pinverse once these
```

Comment:
regarding the todo comment, `torch.qr` should always be preferable to `torch.svd`, being much (probably ~10x) faster and comparably stable. `torch.pinverse` currently uses `torch.svd` internally, but this is a bad thing and it should use `qr`.